Python support - quickstart (#2173)
openvinotoolkit/model_server: docs/python_support/quickstart.md

# Python support in OVMS - quickstart {#ovms_docs_python_support_quickstart}

OpenVINO Model Server allows users to write custom processing nodes in Python, so they may have full control over what happens with the data reaching such node and what comes out of it. Learn more about [Python support](). 

In this quickstart you'll create a servable with a single custom Python code that will expect a string and return the same string, but all in uppercase. 

To acheive that let's follow the steps:
1. Prepare Docker Image And Workspace
2. Prepare Python Code For The Server And Client
3. Prepare Graph Configuration File
4. Prepare Server Configuration File
5. Run Model Server
6. Send Request

### Step 1: Prepare Docker Image And Workspace

Since changing all the letters in the string to uppercase is a very basic use case, the basic Python-enabled model server image is enough. If you'd need some external modules (like opencv), you would need specifically add them. 
In this quickstart it's not required, so just get the basic one:

```

> docker pull openvino/model_server:latest

Reviewer: This is not correct while the image should be built with special params
public image will have in the future different name probably with full tag

```

Also create a new working directory:
```bash
mkdir workspace && cd workspace
```

> You'll also need a client module so in your environment run:

Reviewer: I suggest to separate client and server configuration

```bash
echo '
input_stream: "OVMS_PY_TENSOR:text"
output_stream: "OVMS_PY_TENSOR:uppercase"

node {
  name: "pythonNode"
  calculator: "PythonExecutorCalculator"
  input_side_packet: "PYTHON_NODE_RESOURCES:py"
  input_stream: "INPUT:text"
  output_stream: "OUTPUT:uppercase"
  node_options: {
    [type.googleapis.com / mediapipe.PythonExecutorCalculatorOptions]: {
      handler_path: "/servable/model.py"
    }
  }
}' >> servable/graph.pbtxt
```

Above configuration file creates a graph with a single Python node that uses `PythonExecutorCalculator`, set's inputs and outputs and provides your Python code location in `handler_path`. 
The `input_side_packet` value is an internal field used by the model server to share resources between graph instances - do not change it. 

### Step 4: Prepare Server Configuration File

Last piece of configuration would be the model server configuration file. 
In the `servable` catalog create `config.json` with the following content:

```bash
echo '
{
    "model_config_list": [],
    "mediapipe_config_list": [
    {
        "name":"python_model",
        "graph_path":"/servable/graph.pbtxt"
    }
    ]
}
' >> servable/config.json

> ```

Reviewer: maybe showing the structure of workspace would be helpful

### Step 5: Run Model Server

Now let's run the server:
```bash

> docker run -it --rm -p 11339:11339 -v $PWD/servable:/servable openvino/model_server:latest --config_path /servable/config.json --port 11339

Reviewer: standard port 9000?

# Python support in OVMS - quickstart {#ovms_docs_python_support_quickstart}

OpenVINO Model Server allows users to write custom processing nodes in Python, so they may have full control over what happens with the data reaching such node and what comes out of it.

In this quickstart you'll create a servable with a single custom Python code that will expect a string and return the same string, but all in uppercase. 

> To acheive that let's follow the steps:

Reviewer 2: In sphinx, it's possible to create links that will follow you to specific part of the document. We already do that in some docs. Can we use that here?

Author: This may appear on the side menu like for main quickstart:
https://docs.openvino.ai/2023.2/ovms_docs_quick_start_guide.html#doxid-ovms-docs-quick-start-guide
I'll confirm that, but if it is, then I think it's obsolete

1. Prepare Docker Image And Workspace
2. Create Client Application
3. Write Python Code For The Server 
4. Prepare Graph Configuration File
5. Prepare Server Configuration File
6. Deploy OpenVINO Model Server
7. Send Requests From The Client

### Step 1: Prepare Docker Image And Workspace

> Let's have all the work done in a new `workspace` directory. Also let's  the following subdirectories:

Reviewer 2: ```suggestion
Separate all required files in new `workspace` directory. Also, create following subdirectories:
```

- `workspace/models` (catalog that will be mounted to the deployment)
- `workspace/models/python_model` (catalog with Python servable specification)

You can do that in one go:

```bash
mkdir -p workspace/models/python_model && cd workspace
```

> Since changing all the letters in the string to uppercase is a very basic use case, the basic Python-enabled model server image without extra Python packages is sufficient. If you'd need some external modules, you would need specifically add them to the image. 

Reviewer 2: ```suggestion
Since changing all the letters in the string to uppercase is a very basic example, the basic Python-enabled model server image without extra Python packages is sufficient. If you need some external modules, add them to the image manually. 
```
Should we use shortcuts in our doc such as you'll/you'd etc? I think we avoided it up to this point to be more formal. I remember also the feedback from external team that we should avoid refering to the reader directly.

Until the image is published in next release, the image needs to be built from sources:

```bash
git clone https://github.com/openvinotoolkit/model_server.git
cd model_server
make docker_build MEDIAPIPE_DISABLE=0 PYTHON_DISABLE=0 OV_USE_BINARY=1 RUN_TESTS=0
cd ..
```

You'll also need a client module so in your environment install a required dependency:
```bash
pip3 install tritonclient[grpc]
```

### Step 2: Create Client Application

> Server provides the data to your Python node as-is, so in the server side Python code you can expect an untouched data buffer and some metadata from the request. This means that your client and server code need to be complementary.

Reviewer 2: At first it was had for me to understand the first sentence. Maybe explain buffer protocol?

Author: I wanted to hide the details at this point. Is it better now?

The last part would be to send this data to the server:

```python
results = client.infer("python_model", [infer_input])
print(results.as_numpy("OUTPUT").tobytes().decode())
```

That part will pack `infer_input` into a request and send it to the servable called `uppercase_model`. In the second line it reads output called `OUTPUT` from the response and decodes it to string and prints it. Note that we expect to receive the string in a specific format. In this case, also UTF-8 encoded, so the server side code will need to provide the result in that format.

Let's save the entire code to `client.py` file inside `workspace`:

```bash
echo '
import tritonclient.grpc as grpcclient
client = grpcclient.InferenceServerClient("localhost:9000")
data = "Make this text uppercase.".encode()
infer_input = grpcclient.InferInput("text", [len(data)], "BYTES")
infer_input._raw_content = data
results = client.infer("python_model", [infer_input])
print(results.as_numpy("OUTPUT").tobytes().decode())
' >> client.py 
```

### Step 3: Write Python Code For The Server 

> Moving on to the server side code. The server will provide inputs in a form of `list` of `pyovms.Tensor` objects. In this case you'll have only one input so the code can start like this:

Author: I didn't want to include any code that is not relevant to the presented functionality. Such information could be linked here as soon as we have full documentation.

```bash
echo '
input_stream: "OVMS_PY_TENSOR:text"
output_stream: "OVMS_PY_TENSOR:uppercase"

node {
  name: "pythonNode"
  calculator: "PythonExecutorCalculator"
  input_side_packet: "PYTHON_NODE_RESOURCES:py"
  input_stream: "INPUT:text"
  output_stream: "OUTPUT:uppercase"
  node_options: {
    [type.googleapis.com / mediapipe.PythonExecutorCalculatorOptions]: {
      handler_path: "/models/python_model/model.py"
    }
  }
}' >> models/python_model/graph.pbtxt
```

Above configuration file creates a graph with a single Python node that uses `PythonExecutorCalculator`, set's inputs and outputs and provides your Python code location in `handler_path`. 
The `input_side_packet` value is an internal field used by the model server to share resources between graph instances - do not change it. 

### Step 5: Prepare Server Configuration File

Last piece of configuration would be the model server configuration file. 

> Create `config.json` with the following content in `workspace/models`:

Reviewer 2: ```suggestion
Create `config.json` with the following content in `workspace/models`.
This tells OVMS to to serve the graph under given name `python_model`.
```

```bash
echo '
{
    "model_config_list": [],
    "mediapipe_config_list": [
    {
        "name":"python_model",
        "graph_path":"/models/python_model/graph.pbtxt"
    }
    ]
}
' >> models/config.json
```

### Step 6: Deploy OpenVINO Model Server

Before running the server let's check if you have all the required files. You can do that with a `tree` command:
```bash
tree
.
├── client.py
└── models
    ├── config.json
    └── python_model
        ├── graph.pbtxt
        └── model.py
```

Now let's run the server:
```bash
docker run -it --rm -p 9000:9000 -v $PWD/models:/models openvino/model_server:latest --config_path /models/config.json --port 9000
```
### Step 7: Send Requests From The Client

Once you have model server up and running, let's send a text: `"Make this text uppercase."`. 

Simply run your `client.py` from the `workspace` catalog and see the results:

```bash
python3 client.py
MAKE THIS TEXT UPPERCASE.
```

> 

Reviewer 2: Should we add this page as a link to some other page?

Author: I'll link it

To achieve that let's follow the steps:
1. Prepare Docker Image And Workspace
2. Write Python Code For The Server 
3. Prepare Graph Configuration File
4. Prepare Server Configuration File
5. Deploy OpenVINO Model Server
6. Create Client Application
7. Send Requests From The Client

### Step 1: Prepare Docker Image And Workspace

Let's have all the work done in a new `workspace` directory. Also create the following subdirectories:
- `workspace/models` (catalog that will be mounted to the deployment)
- `workspace/models/python_model` (catalog with Python servable specification)

You can do that in one go:

```bash
mkdir -p workspace/models/python_model && cd workspace
```

Since changing all the letters in the string to uppercase is a very basic example, the basic Python-enabled model server image without extra Python packages is sufficient. If you need some external modules, you need to add them to the image manually. 
Until the image is published in next release, it needs to be built from sources:

```bash
git clone https://github.com/openvinotoolkit/model_server.git
cd model_server
make docker_build MEDIAPIPE_DISABLE=0 PYTHON_DISABLE=0 OV_USE_BINARY=1 RUN_TESTS=0
cd ..
```

> You will also need a client module so in your environment install a required dependency:

Reviewer 3: ```suggestion
You will also need a client module, so in your environment install a required dependency:
```

```bash
pip3 install tritonclient[grpc]
```

### Step 2: Write Python Code For The Server 

Let's start with the server side code. Your job is to implement an `OvmsPythonModel` class. Model Server expects it to have at least `initialize` and `execute` methods.

> Since there's nothing to initialize, let's focus on `execute` method that will be called everytime model receives a request. The server will read inputs from that request and pass them to `execute` function as an `inputs` argument. 

Reviewer 3: ```suggestion
Since there's nothing to initialize, let's focus on `execute` method that will be called every time model receives a request. The server will read inputs from that request and pass them to `execute` function as an `inputs` argument. 
```

`inputs` is a `list` of `pyovms.Tensor` objects. In this case you will have only one input so the code can start like this:

```python
input_data = inputs[0]
```

Now `input_data` is `pyovms.Tensor` object that holds the data and some metadata like shape and datatype. At this point you need to decide what kind of data you expect to receive here.

> You will work on a string, so let's say you expect `input_data` to be UTF-8 encoded string. In that case you can create an instances of `bytes` from `input_data` and then decode it to actual string:

Reviewer 3: ```suggestion
You will work on a string, so let's say you expect `input_data` to be UTF-8 encoded string. In that case you can create an instances of `bytes` from `input_data` and then decode it to an actual string:
```

```bash
echo '
from pyovms import Tensor

class OvmsPythonModel:

    def initialize(self, kwargs: dict):
        return True

    def execute(self, inputs: list):
        input_data = inputs[0]
        text = bytes(input_data).decode()
        output_data = text.upper().encode()
        return [Tensor("OUTPUT", output_data)]
' >> models/python_model/model.py
```

### Step 3: Prepare Graph Configuration File

Python logic execution in OVMS is supported via Mediapipe graphs. That means you will need to prepare graph definition for your processing flow. In that case, a graph with just one node - Python node - is enough. Let's create appropriate `graph.pbtxt` file in your `workspace/models/python_model` catalog:

```bash
echo '
input_stream: "OVMS_PY_TENSOR:text"
output_stream: "OVMS_PY_TENSOR:uppercase"

node {
  name: "pythonNode"
  calculator: "PythonExecutorCalculator"
  input_side_packet: "PYTHON_NODE_RESOURCES:py"
  input_stream: "INPUT:text"
  output_stream: "OUTPUT:uppercase"
  node_options: {
    [type.googleapis.com / mediapipe.PythonExecutorCalculatorOptions]: {
      handler_path: "/models/python_model/model.py"
    }
  }
}' >> models/python_model/graph.pbtxt
```

> Above configuration file creates a graph with a single Python node that uses `PythonExecutorCalculator`, set's inputs and outputs and provides your Python code location in `handler_path`. 

Reviewer 3: ```suggestion
Above configuration file creates a graph with a single Python node that uses `PythonExecutorCalculator`, sets inputs and outputs and provides your Python code location in `handler_path`. 
```

The `input_side_packet` value is an internal field used by the model server to share resources between graph instances - do not change it. 

### Step 4: Prepare Server Configuration File

Last piece of configuration would be the model server configuration file. 
Create `config.json` with the following content in `workspace/models`:

```bash
echo '
{
    "model_config_list": [],
    "mediapipe_config_list": [
    {
        "name":"python_model",
        "graph_path":"/models/python_model/graph.pbtxt"
    }
    ]
}
' >> models/config.json
```

This tells OVMS to to serve the graph under given name `python_model`.

### Step 5: Deploy OpenVINO Model Server

Before running the server let's check if all files required for deployment are in place. Check the contents of `workspace/models` catalog as it will be mounted to the container:

> ```

Reviewer 3: ```suggestion
```bash
```

tree models
models
├── config.json
└── python_model
    ├── graph.pbtxt
    └── model.py
```

Now let's run the server:
```bash
docker run -it --rm -p 9000:9000 -v $PWD/models:/models openvino/model_server:latest --config_path /models/config.json --port 9000
```

### Step 6: Create Client Application

Now that the Python model is deployed, you can focus on the other end - the client application. When writing the client keep in mind how the server side code looks like as they must be complementary.

First let's connect to the server hosted on `localhost` with gRPC interface available on port `9000`:

```python
import tritonclient.grpc as grpcclient
client = grpcclient.InferenceServerClient("localhost:9000")
```

You will send a string, so let's create one and encode it to UTF-8, because that's what the server side code expects:

```python
data = "Make this text uppercase.".encode()
```

Now let's pack that data into a gRPC structure that will be sent to the server:

```python
infer_input = grpcclient.InferInput("text", [len(data)], "BYTES")
infer_input._raw_content = data

> ```

Reviewer 3: ```suggestion
```

```

> You've created InferInput object that will correspond to the servable input with name "text" (more about it soon), shape of (number_of_encoded_bytes) and datatype "BYTES". The data itself has been written to raw_content field. All of these values can be accessed on the server side.

Reviewer 3: ```suggestion

You've created InferInput object that will correspond to the servable input with the name "text" (more about it soon), shape of (number_of_encoded_bytes) and datatype "BYTES". The data itself has been written to a raw_content field. All of these values can be accessed on the server side.
```

> You've created InferInput object that will correspond to the servable input with name "text" (more about it soon), shape of (number_of_encoded_bytes) and datatype "BYTES". The data itself has been written to raw_content field. All of these values can be accessed on the server side.

Reviewer 3: "shape of (number_of_encoded_bytes)" - is it missing something? or "of" is not required

> You've created InferInput object that will correspond to the servable input with the name "text" (more about it soon), shape of (number_of_encoded_bytes) and datatype "BYTES". The data itself has been written to a raw_content field. All of these values can be accessed on the server side.

Author: ```suggestion
You've created InferInput object that will correspond to the servable input with the name "text" (more about it soon), shape [len(data)] - where len(data) is the number of encoded bytes - and datatype "BYTES". The data itself has been written to a raw_content field. All of these values can be accessed on the server side.
```

The last part would be to send this data to the server:

```python
results = client.infer("python_model", [infer_input])
print(results.as_numpy("OUTPUT").tobytes().decode())
```

That part will pack `infer_input` into a request and send it to the servable called `uppercase_model`. 

Server is expected to respond with an output containing UTF-8 encoded string, so in the second line you read it, decode it to actual string and print it.

Let's save the entire code to `client.py` file inside `workspace`:

```bash
echo '
import tritonclient.grpc as grpcclient
client = grpcclient.InferenceServerClient("localhost:9000")
data = "Make this text uppercase.".encode()
infer_input = grpcclient.InferInput("text", [len(data)], "BYTES")
infer_input._raw_content = data
results = client.infer("python_model", [infer_input])
print(results.as_numpy("OUTPUT").tobytes().decode())
' >> client.py 
```

> 

Reviewer 3: ```suggestion
```
